Subject: x86: activate HARDIRQS_SW_RESEND
From: Ingo Molnar <mingo@elte.hu>

activate the software-triggered IRQ-resend logic.

it appears some chipsets/cpus do not handle local-APIC driven IRQ
resends all that well, so always use the soft mechanism to trigger
the execution of pending interrupts.

Signed-off-by: Ingo Molnar <mingo@elte.hu>
---
 arch/x86/Kconfig    |    4 ++++
 kernel/irq/manage.c |    8 ++++++++
 2 files changed, 12 insertions(+)

Index: linux-2.6.24.7-rt21/arch/x86/Kconfig
===================================================================
--- linux-2.6.24.7-rt21.orig/arch/x86/Kconfig	2008-10-08 22:23:40.000000000 -0400
+++ linux-2.6.24.7-rt21/arch/x86/Kconfig	2008-10-08 22:24:25.000000000 -0400
@@ -1230,6 +1230,10 @@ config OUT_OF_LINE_PFN_TO_PAGE
 	def_bool X86_64
 	depends on DISCONTIGMEM
 
+config HARDIRQS_SW_RESEND
+	bool
+	default y
+
 menu "Power management options"
 	depends on !X86_VOYAGER
 
Index: linux-2.6.24.7-rt21/kernel/irq/manage.c
===================================================================
--- linux-2.6.24.7-rt21.orig/kernel/irq/manage.c	2008-10-08 22:24:21.000000000 -0400
+++ linux-2.6.24.7-rt21/kernel/irq/manage.c	2008-10-08 22:24:25.000000000 -0400
@@ -191,6 +191,14 @@ void enable_irq(unsigned int irq)
 		desc->depth--;
 	}
 	spin_unlock_irqrestore(&desc->lock, flags);
+#ifdef CONFIG_HARDIRQS_SW_RESEND
+	/*
+	 * Do a bh disable/enable pair to trigger any pending
+	 * irq resend logic:
+	 */
+	local_bh_disable();
+	local_bh_enable();
+#endif
 }
 EXPORT_SYMBOL(enable_irq);
 
